### DISCUSSION TITLE
Including the class name in serialization, "Already mutably borrowed"

### DISCUSSION THREAD

**Discussion by bradbeattie** [pydantic/pydantic#6538] [Question]
So I'd like to be able to serialize my models with the class name. Migration docs point me towards [@model_serializer](https://docs.pydantic.dev/latest/usage/serialization/#custom-serializers).

However, when I give that a try with nested dicts like so...

```python
class Foo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    x: dict[str, ForwardRef("Foo") | str]
    
    @model_serializer
    def ser_model(self) -> dict[str, Any]:
        return {**dict(self), "_class": self.__class__.__name__}
        # Foo in practice is a new base class, so I can't just return `{"x": self.x, "_class": "Foo"}`
    

myfoo = Foo(x={
    "first": Foo(x={
        "second": "foobar"
    })
})
```

* 🟢 `print(myfoo.model_dump())` prints `{'x': {'first': {'x': {'second': 'foobar'}, '_class': 'Foo'}}, '_class': 'Foo'}`
* 🟥 `print(myfoo.model_dump_json())` fails with `Error serializing to JSON: RuntimeError: Already mutably borrowed` (Rust, I suspect)

I suppose I can `print(json.dumps(myfoo.model_dump()))`, but that feels like a hacky workaround. Am I doing something wrong here or is this a legitimate bug I should file?

**Comment by bradbeattie** [comment]
Working around this with

```python
class Foo(BaseModel):
    model_config = ConfigDict(extra="forbid")
    class_name: str

    def __init__(self, **kwargs):
        kwargs["class_name"] = self.__class__.__name__
        super().__init__(**kwargs)
```

**Reply by wv-opt-ai** [reply]
Good idea. I build on your example with this way using pydantic's decorators:
```python
from pydantic import BaseModel, Field, field_validator


class Foo(BaseModel):
    class_name: str | None = Field(
        default=None,
        description="Automated - do not set.",
        validate_default=True,  # important to enable the injector!
    )

    @field_validator("class_name", mode="before")
    @classmethod
    def inject_class_name(cls, value: str) -> str:  # noqa: ARG003
        """Inject the name, ignoring the input."""
        return cls.__name__
```

so it gets automatically injected without overriding the `__init__` method:
```
> Foo().class_name
"Foo"
```
